actors/builtin/market/deal.go

```diff
@@ -12,6 +24,184 @@ import (
 //}
 var PieceCIDPrefix = market0.PieceCIDPrefix
 
+// The DealLabel is a kinded union of string or byte slice.
+// It serializes to a CBOR string or CBOR byte string depending on which form it takes.
+// The empty label is serialized as an empty CBOR string (maj type 3).
+type DealLabel struct {
+	s  *string
+	bs *[]byte
+}
```

Comment:
personal opinion: I'd write this as

```
type DealLabel struct {
	bs []byte
	notString bool
}
```

Your version is more "go-like", but also requires extra allocations and may be a bit annoying to deal with...

Reply:
Ironically I was following up on your comment [here](https://github.com/filecoin-project/specs-actors/pull/1580#discussion_r826315784) is this not the "put them behind pointers" approach?

Reply:
I'm going to move it to the above approach

Reply:
> Ironically I was following up on your comment https://github.com/filecoin-project/specs-actors/pull/1580#discussion_r826315784 is this not the "put them behind pointers" approach?

Your version is the "put them behind pointers approach". My version here is the "enum thing" approach (which, after looking at this, seems simpler.

> nvm notString is the way it is to make this nice

Welcome to go! zero values are over there :point_right: .